app/src/main/java/org/wikipedia/diff/ArticleEditDetailsFragment.kt

```diff
@@ -0,0 +1,415 @@
+        view.setTextColor(themedColor)
+        view.iconTint = (ColorStateList.valueOf(themedColor))
+    }
+
+    private fun watchOrUnwatchTitle(@Nullable expiry: WatchlistExpiry?, unwatch: Boolean) {
+        disposables.add(ServiceFactory.get(WikiSite.forLanguageCode(languageCode)).watchToken
+                .subscribeOn(Schedulers.io())
+                .flatMap { response: MwQueryResponse ->
+                    val watchToken = response.query()!!.watchToken()
+                    if (TextUtils.isEmpty(watchToken)) {
+                        throw RuntimeException("Received empty watch token: " + GsonUtil.getDefaultGson().toJson(response))
+                    }
+                    ServiceFactory.get(WikiSite.forLanguageCode(languageCode)).postWatch(if (unwatch) 1 else null, null, articleTitle,
+                            expiry?.expiry, watchToken!!)
+                }
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe({ watchPostResponse: WatchPostResponse ->
+                    val firstWatch = watchPostResponse.getFirst()
+                    isWatched = firstWatch!!.watched
+                    updateWatchlistButtonUI()
+                    // Reset to make the "Change" button visible.
+                    if (watchlistExpiryChanged && unwatch) {
+                        watchlistExpiryChanged = false
+                    }
+                    showWatchlistSnackbar(expiry, firstWatch)
+                }) { t: Throwable? -> L.d(t) })
+    }
+
+    private fun updateWatchlistButtonUI() {
+        setButtonTextAndIconColor(watchButton, ResourceUtil.getThemedColor(requireContext(),
+                if (isWatched) R.attr.color_group_62 else R.attr.colorAccent))
+        watchButton.text = getString(if (isWatched) R.string.watchlist_details_watching_label
+        else R.string.watchlist_details_watch_label)
+    }
+
+    private fun showWatchlistSnackbar(@Nullable expiry: WatchlistExpiry?, watch: Watch) {
+        isWatched = watch.watched
+        if (watch.unwatched) {
+            FeedbackUtil.showMessage(this, getString(R.string.watchlist_page_removed_from_watchlist_snackbar, articleTitle))
+            watchlistExpirySession = null
+
+        } else if (watch.watched && expiry != null) {
+            val snackbar = FeedbackUtil.makeSnackbar(requireActivity(),
+                    getString(R.string.watchlist_page_add_to_watchlist_snackbar,
+                            articleTitle,
+                            getString(expiry.stringId)),
+                    FeedbackUtil.LENGTH_DEFAULT)
+            if (!watchlistExpiryChanged) {
+                snackbar.setAction(R.string.watchlist_page_add_to_watchlist_snackbar_action) {
+                    watchlistExpiryChanged = true
+                    bottomSheetPresenter.show(childFragmentManager, WatchlistExpiryDialog.newInstance(expiry))
+                }
+            }
+            watchlistExpirySession = null
+            snackbar.show()
```

Comment:
Looks like the snackbar does not show properly. Also, the "watching" button should be disabled when posting the request and enable until the process and the status is received.